fl4health/preprocessing/autoencoders/dim_reduction.py

```diff
@@ -0,0 +1,84 @@
+from abc import ABC
+from pathlib import Path
+
+import torch
+
+
+class Processing(ABC):
+    """
+    Abstact class to add processing methods.
+    """
+
+    def __init__(
+        self,
+        checkpointing_path: Path,
+    ) -> None:
+        self.checkpointing_path = checkpointing_path
+        self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        self.load_autoencoder()
+
+    def load_autoencoder(self) -> None:
+        autoencoder = torch.load(self.checkpointing_path)
+        autoencoder.eval()
+        self.autoencoder = autoencoder.to(self.device)
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}"
+
+
+class AeProcessor(Processing):
+    """Transformer processor to encode the data using basic autoencoder."""
+
+    def __init__(
+        self,
+        checkpointing_path: Path,
+    ) -> None:
+        super().__init__(checkpointing_path)
+
+    def __call__(self, sample: torch.Tensor) -> torch.Tensor:
+        # This transformer is called for the input samples after they are transfered into torch tensors.
+        embedding_vector = self.autoencoder.encode(sample.to(self.device))
+        return embedding_vector.clone().detach()
+
+
+class VaeProcessor(Processing):
+    """Transformer processor to encode the data using VAE encoder."""
+
+    def __init__(
+        self,
+        checkpointing_path: Path,
+        include_variance: bool = False,
+    ) -> None:
+        super().__init__(checkpointing_path)
+        self.include_variance = include_variance
+
+    def __call__(self, sample: torch.Tensor) -> torch.Tensor:
+        # This transformer is called for the input samples after they are transfered into torch tensors.
+        mu, logvar = self.autoencoder.encode(sample.to(self.device))
+        if self.include_variance:
+            return torch.cat(mu.clone().detach(), logvar.clone().detach())
+        return mu.clone().detach()
+
+
+class CvaeFixedConditionProcessor(Processing):
```

Comment:
We should probably have an implementation of a "non-fixed" conditional processor, at the very least for demonstration purposes, but also for practical settings like label based processing etc.?

Reply:
Previously, there was a label-based processing in this file, but we removed it since we assumed during inference at test time we don't have access to the labels. It was this part of the code:

class LabelConditionedProcessor(Processing):
"""Transformer processor to encode the data using CVAE encoder, with the labels as conditions."""

    def __init__(
        self,
        checkpoint_path: Path,
    ) -> None:
        super().__init__(checkpoint_path)

    def __call__(self, sample: np.ndarray, target: np.ndarray) -> Tuple[torch.Tensor, torch.Tensor]:
        condition_vector = self.autoencoder.one_hot(torch.tensor(target))
        mu, logvar = self.autoencoder.encode(sample, condition_vector)
        return mu, `torch.from_numpy(target`
 I think for demonstration purposes we can add something similar to this (considering the new changes).

Reply:
For this comment/idea, let's put a ticket in the backlog to add rather than address it now.

Reply:
I added `CvaeNonFixedConditionProcessor` and it is tested in `tests/preprocessing/test_ae_dim_reduction.py`